File: Algorithm/Python/953. Verifying an Alien Dictionary.py
# -*- coding: utf-8 -*-
# Created on Fri Dec 14 2018 14:15:15
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# compare char by char, pay attention to the case when one word is the prefix of another word
class Solution(object):
    def isAlienSorted(self, words, order):
        """
        :type words: List[str]
        :type order: str
        :rtype: bool
        """
        order_num = {}
        for i, c in enumerate(order):
            order_num[c] = i
        for i in xrange(len(words) - 1):
            equal = True
            for j in xrange(min(len(words[i]), len(words[i+1]))):
                if order_num[words[i][j]] < order_num[words[i+1][j]]:
                    equal = False
                    break
                elif order_num[words[i][j]] > order_num[words[i+1][j]]:
                    return False
            if equal and len(words[i]) > len(words[i+1]):
                return False
        return True

File: Algorithm/C++/953. Verifying an Alien Dictionary.cc
/*
 * Created on Fri Dec 14 2018 14:35:58
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// compare char by char, pay attention to the case when one word is the prefix of another word
#include <vector>
#include <string>
#include <unordered_map>

using std::vector;
using std::string;
using std::unordered_map;

class Solution {
 public:
  bool isAlienSorted(vector<string>& words, string order) {
    unordered_map<char, int> order_num;
    for(int i = 0; i < order.size(); i++) {
      order_num[order[i]] = i;
    }

    for (int i = 0; i < words.size() - 1; i++) {
      bool equal = true;
      for (int j = 0; j < std::min(words[i].size(), words[i+1].size()); j++) {
        if(order_num[words[i][j]] < order_num[words[i+1][j]]) {
          equal = false;
          break;
        } else if(order_num[words[i][j]] > order_num[words[i+1][j]]) {
          return false;
        }
      }
      if (equal && words[i].size() > words[i+1].size()) {
        return false;
      }
    }
    return true;
  }
};